lib/webpacker/manifest.rb

```diff
@@ -13,32 +13,94 @@ def file_path
+      instance.confirm_manifest_exists
+
       if Webpacker::Configuration.compile?
-        compile_and_find!(name)
+        compile_and_find(name, throw_if_missing: throw_if_missing)
       else
-        find!(name)
+        # Since load_instance checks a `mtime` on the manifest for a non-production env before loading,
+        # we should always call this before a call to `find!` since one may be using
+        # `webpack -w` and a file may have been added to the manifest since Rails first started.
+        load_instance
+        raise Webpacker::FileLoader::FileLoaderError.new("Webpacker::Manifest.load must be called first") unless instance
+
+        return find(name, throw_if_missing: throw_if_missing)
       end
     end
 
+    # Why does this method exist? Testing? It's not in the README
     def lookup_path(name)
-      Rails.root.join(File.join(Webpacker::Configuration.public_path, lookup(name)))
+      Rails.root.join(File.join(Webpacker::Configuration.output_path, lookup(name)))
+    end
+
+    # Helper method to determine if the manifest file exists. Maybe Webpack needs to run?
+    # **Used by React on Rails.**
+    def exist?
+      path_object = Webpacker::Configuration.manifest_path
+      path_object.exist?
     end
 
     private
-      def find!(name)
-        raise Webpacker::FileLoader::FileLoaderError.new("Webpacker::Manifest.load must be called first") unless instance
-        instance.data[name.to_s] || raise(Webpacker::FileLoader::NotFoundError.new("Can't find #{name} in #{file_path}. Is webpack still compiling?"))
+      def missing_file_from_manifest_error(bundle_name)
+        msg = <<-MSG
+Webpacker can't find #{bundle_name} in your manifest at #{file_path}. Possible causes:
+  1. You are hot reloading.
+  2. You want to set Configuration.compile to true for your environment.
+  3. Webpack has not re-run to reflect updates.
+  4. You have misconfigured Webpacker's config/webpacker.yml file.
+  5. Your Webpack configuration is not creating a manifest.
+Your manifest contains:
+#{instance.data.to_json}
+        MSG
+        raise(Webpacker::FileLoader::NotFoundError.new(msg))
       end
 
-      def compile_and_find!(name)
+      def missing_manifest_file_error(path_object)
+        msg = <<-MSG
+Webpacker can't find the manifest file: #{path_object}
+Possible causes:
+  1. You have not invoked webpack.
+  2. You have misconfigured Webpacker's config/webpacker_.yml file.
+  3. Your Webpack configuration is not creating a manifest.
+MSG
+        raise(Webpacker::FileLoader::NotFoundError.new(msg))
+      end
+
+      def find(name, throw_if_missing: true)
+        value = instance.data[name.to_s]
+        return value if value.present?
+
+        if throw_if_missing
+          missing_file_from_manifest_error(name)
+        end
+      end
+
+      def compile_and_find!(name, throw_if_missing: true)
         Webpacker.compile
-        find!(name)
+        find(name, throw_if_missing: throw_if_missing)
```

Comment:
Hi @dhh and @gauravtiwari,

1. Nobody (as far as I know) is ever using "watch" in production.
2. We do use "watch" for development and test files. That is the default from the React on Rails generator. We need the ability to _not_ throw if a file is missing so that Webpack can be _generically_ invoked to compile assets.